### PR TITLE
breaking changes to MLJGLMInterface.jl

### PR DESCRIPTION
**Pull request by OkonSamuel**
This PR includes quite a number of breaking and non-breaking changes. These changes include:
1. Remove `Parameters.jl` dependency: This dependency is no longer required as the macro `@with_kw_noshow`  is also been removed in this PR. The `@with_kw_noshow` macro has been replace with `@mlj_model` macro. This PR also get's rid of type parameters in the model structs, which could affect hyperparameter tuning of these models as type parameters (although helpful in performance) restrict the range of objects an hyper-parameter could take. 
2. Introduces a new fit return object `FitResult` which houses the glm model coefficients (including intercept), dispersion parameter of the glm distribution, and any other fit result that may be specific to a given glm model. This has the advantage that it no longer houses the fit data, hence allowing serialization of only the key components of a model.
3. Switch from formula call to matrix: Prior to this PR `MLJGLMInterface` `fit` method at its core makes a call to `glm(formula, data, distr::UnivariateDistribution, link::Link = canonicallink(d); <keyword arguments>)` where `data` is a table containing the features and also includes the target column. This PR replaces this call with  ` glm(X::AbstractMatrix, y::AbstractVector, distr::UnivariateDistribution, link::Link = canonicallink(d); <keyword arguments>)` where `X` is the feature matrix(with a column of ones appended at the end if `intercept` is being fitted), and `y` is the the target vector. This has the advantage of avoiding unnecessary costly allocations that result from manually having to create a table that includes features only to split such a table latter on.
4. added a new internal method `_to_matrix` which is a more efficient way of converting a Table to a matrix and also adjusting the matrix when the model also fits an intercept.
5. Added extra checks on the data which catches bugs there were previously waiting to be uncovered. For example for the case where one trains a glm model with an offset column set (i.e `model.offsetcol !== nothing`), `fit_intercept==false` and a dataset containing one feature and one target, wouldn't make any sense and this PR throws (hopefully) and informative error message as to why. Also fitting glm models on dataset containing more features than observations would lead to incorrect results, so checks are in place to throw errors in this case.
6. Allow users to customize the `report`: This PR also allows the user to customize the report with the `report_parms` model hyper-parameter, there are several options available  for example setting `report_params = nothing` which returns an empty `NamedTuple` (indicating no report/statistic was computed). `report_params` can also be an instance of `Vector{Symbol}` with elements being members of `[:deviance, :dof_residual, :stderror, :vcov, :coef_table]`. By default all possible options are computed in the report. 
7. As a result of (2) and (6) above, the nice nice view of the cool glm coefficient table introduced in https://github.com/JuliaAI/MLJGLMInterface.jl/pull/12, is no longer the display of the `fitresult`. This has been instead been moved to one of the optional items to be added to the `report`.
8. The PR also fixes a bug in `LinearRegressor` model: This model accepts the keyword argument `allowrankdeficient` which was never used during the fit. So this PR incorporates fixes this by replacing this kwarg with the more recent `dropcollinear` (Derived from GLM package) and taking account of this keyword argument during the fit by replacing the `GLM.glm` method with `GLM.lm` method.
9. Finally this PR adds supports for observation weights to all implemented GLM models. These weight are directly passed to the GLM models. According to the GLM package documentation 
> Such weights are equivalent to repeating each observation a number of times equal to its weight. Do note that this interpretation gives equal point estimates but different standard errors from analytical (a.k.a. inverse variance) weights and from probability (a.k.a. sampling) weights which are the default in some other software.

cc. @ablaom , @rikhuijzer.

### CODE REVIEW

**Comment by rikhuijzer** [comment]
Ok, so as a reviewer this is a very difficult PR to review. There is too much changing at the same time which makes it hard to tell what are the implications of the changes. Could we maybe split up this PR into separate PRs?

Then, regarding the changes listed in the description: Point 1 about dropping the `Parameters.jl` dependency seems fine. Point 2  about the new `FitResult` which doesn't contain the data is related to #16 I guess? Point 3 about dropping `@formula` support is something I don't get yet. This package is called MLJGLMInterface, so having an interface to the formula syntax seems like the main use-case for this package. For performance, shouldn't people use https://github.com/JuliaAI/MLJLinearModels.jl? Point 7 is about not having the report feature names in fitresult. This is caused by dropping the `@formula` right? Point 8 and 9 seem like good fixes.

**Comment by OkonSamuel** [comment]
> Ok, so as a reviewer this is a very difficult PR to review. There is too much changing at the same time which makes it hard to tell what are the implications of the changes. Could we maybe split up this PR into separate PRs?
> 
> Then, regarding the changes listed in the description: Point 1 about dropping the `Parameters.jl` dependency seems fine. Point 2 about the new `FitResult` which doesn't contain the data is related to #16 I guess? Point 3 about dropping `@formula` support is something I don't get yet. This package is called MLJGLMInterface, so having an interface to the formula syntax seems like the main use-case for this package. For performance, shouldn't people use https://github.com/JuliaAI/MLJLinearModels.jl? Point 7 is about not having the report feature names in fitresult. This is caused by dropping the `@formula` right? Point 8 and 9 seem like good fixes.

Thanks for taking a look  at the PR. I agree it's kind of difficult to review so I'll instead make a couple of smaller PRs, which hopefully should be easier to review. I'll release them one after the other.

**Comment by OkonSamuel** [comment]
superseded by #19, #20, #21, #22, #23, #24
